idmanager/idmanager.go

```diff
@@ -3,51 +3,71 @@ package idmanager
 import (
 	"context"
 	"fmt"
+	"io"
 
-	"github.com/Scalingo/go-etcd-lock/lock"
+	etcdlock "github.com/Scalingo/go-etcd-lock/lock"
 	"github.com/Scalingo/sand/config"
 	"github.com/Scalingo/sand/etcd"
 	"github.com/Scalingo/sand/store"
 	"github.com/pkg/errors"
 )
 
 type Manager interface {
-	Lock(context.Context) error
+	Lock(context.Context) (Lock, error)
 	Generate(context.Context) (int, error)
-	Unlock(context.Context) error
+}
+
+type Lock interface {
+	Unlock(ctx context.Context) error
 }
 
 type manager struct {
 	store  store.Store
 	config *config.Config
-	lock   lock.Lock
 	field  string
 	name   string
 	prefix string
 }
 
+type lock struct {
+	resourceLock   etcdlock.Lock
+	lockingBackend io.Closer
+}
+
 func New(c *config.Config, s store.Store, name string, field string, prefix string) Manager {
 	return &manager{config: c, store: s, field: field, name: name, prefix: prefix}
 }
 
-func (m *manager) Lock(ctx context.Context) error {
+func (m *manager) Lock(ctx context.Context) (Lock, error) {
+	var l Lock
+
 	client, err := etcd.NewClient()
 	if err != nil {
-		return errors.Wrapf(err, "fail to get etcd client")
+		return l, errors.Wrapf(err, "fail to get etcd client")
 	}
-	l, err := lock.NewEtcdLocker(client).WaitAcquire(fmt.Sprintf("/%s-idgen", m.name), 300)
+	resourceLock, err := etcdlock.NewEtcdLocker(client).WaitAcquire(fmt.Sprintf("/%s-idgen", m.name), 300)
 	if err != nil {
-		return errors.Wrapf(err, "fail to get etcd lock")
+		return l, errors.Wrapf(err, "fail to get etcd lock")
 	}
-	m.lock = l
-	return nil
+	return lock{
+		resourceLock:   resourceLock,
+		lockingBackend: client,
+	}, nil
 }
 
-func (m *manager) Unlock(ctx context.Context) error {
-	if m.lock == nil {
+func (l lock) Unlock(ctx context.Context) error {
+	if l.resourceLock == nil {
 		return errors.New("not locked")
 	}
-	return m.lock.Release()
+	err := l.resourceLock.Release()
+	if err != nil {
+		return errors.Wrapf(err, "fail to release etcd lock")
```

Comment:
Shouldn't you just log the error but keep executing the function to close the `lockingBackend`, even though there has been an error when releasing the `resourceLock`?

Reply:
Fixed